### ISSUE TITLE
Stack too deep in ABIEncoderV2 (ICE in yul::CodeGenerator::assemble(Block &, AsmAnalysisInfo &, evmasm::Assembly &,EVMVersion, ExternalIdentifierAccess &, bool, bool) [Stack too deep when compiling inline assembly: Variable dataEnd is 1 slot(s) too deep inside the stack])

### ISSUE THREAD

**Issue by agroce** [ethereum/solidity#8369]
## Description

This contract:
```solidity
pragma experimental ABIEncoderV2;
contract C{
    struct S{int56 x;}
    function f(int56,S calldata,bytes2,bytes2,bytes2,int56,int56,bytes2,bytes2,bytes2,int56,int56,int56)external{}
}
```

produces:

```
Exception during compilation: /home/user/solidity/libyul/backends/evm/AsmCodeGen.cpp(215): Throw in function static void solidity::yul::CodeGenerator::assemble(const solidity::yul::Block &, solidity::yul::AsmAnalysisInfo &, evmasm::Assembly &, langutil::EVMVersion, const solidity::yul::ExternalIdentifierAccess &, bool, bool)
Dynamic exception type: boost::exception_detail::clone_impl<solidity::yul::YulAssertion>
std::exception::what: Stack too deep when compiling inline assembly: Variable dataEnd is 1 slot(s) too deep inside the stack.
[solidity::util::tag_comment*] = Stack too deep when compiling inline assembly: Variable dataEnd is 1 slot(s) too deep inside the stack.
```

when compiled with solc

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer, now with a few more mutation operators

## Environment

- Compiler version:  0.6.3-develop.2020.2.10+commit.e8eb1f2d.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

**Comment by hrkrshnn** [comment]
It seems that yul function for ABI decoding has additional variables which causes a stack too deep issue.

**Comment by chriseth** [comment]
Stack too deep errors can always happen if there are many function parameters and it can only be detected at code generation phase. @hrkrshnn do you think we should rather create a new exception type that does not translate into an ICE?

**Comment by bshastry** [comment]
@chriseth @hrkrshnn I had filed an issue some time back wishing stack-too-deep errors could be specialized (https://github.com/ethereum/solidity/issues/6772) so that the fuzzer can ignore them. If you think that is still a good idea, I can propose a PR for it. However, I do not think that will help here because solc would still fail but with the newly introduced exception type instead of an ICE.

**Comment by chriseth** [comment]
It is fine to fail, but it should not be considered as bad as an assertion failure.

**Comment by hrkrshnn** [comment]
@chriseth @bshastry I was wondering if it's possible to have a compiler error like the following:

```
Compiler error: Stack too deep, try removing local variables.
 --> /tmp/stack.sol:2:5:
  |
2 |     function f(int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int)external{}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Getting the original location is a non-trivial task, though.

**Comment by chriseth** [comment]
We have an open issue for at least improving it for regular solidity variables.

**Comment by bshastry** [comment]
This issue is "fixed" by #9281
